### ISSUE TITLE
Super call in anon JS class produces invalid IR

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#3055]
Extracted from #3053.

Snippet:
```scala
@ScalaJSDefined
class Foo extends js.Object {
  def bar(msg: String): Unit = println(msg)
}

object HelloWorld extends js.JSApp {
  def main(): Unit = {
    val foo = new Foo {
      override def bar(msg: String): Unit = super.bar("foo: " + msg)
    }
    foo.bar("foobar")
  }
}
```
IR checker error:
```
[error] HelloWorld.scala(20:54:JSSuperBracketCall): JS class type expected but ClassType(Lhelloworld_HelloWorld$$anon$1) found
```

**Comment by sjrd** [comment]
So ... this one is tricky. Not so much because of super calls to *methods*, but because of super calls to *getters and setters*.

Our situation for method calls is as follows. We have a node [`JSSuperBracketCall`](https://github.com/scala-js/scala-js/blob/2350cb6a3c63430f7eb8372a538488ff1c527687/ir/src/main/scala/org/scalajs/core/ir/Trees.scala#L533) (one of our best documented IR nodes ^^) which is used for super calls to methods in JS classes. Example (I omit `@ScalaJSDefined` annotations):
```scala
class Foo extends js.Object {
  def bar(x: Int): Int = x + 1
}

class Bar extends Foo {
  override def bar(x: Int): Int = super.bar(x) * 2
}
```
produces the following IR for `Bar`:
```scala
js class Lhelloworld_Bar extends Lhelloworld_Foo {
  def "constructor"() {
    super()
  }
  def "bar"(arg$1: any): any = {
    val prep0: int = arg$1.asInstanceOf[I];
    Lhelloworld_Bar::bar__Lhelloworld_Bar__I__I(this, prep0)
  }
  static def bar__Lhelloworld_Bar__I__I($this: any, x: int): int = {
    ($this.Lhelloworld_Bar::super["bar"](x).asInstanceOf[I] *[int] 2)
    // ^- this is JSSuperBracketCall(ClassType(Bar), $this, "bar", List(x))
  }
}
```
Because `Bar` is a `js class` that extends `Foo`, the `JSSuperBracketCall` is valid and will lookup `"bar"` in `Foo.prototype`.

If we inline `Bar` as an anonymous class, this is what we get:
```scala
  def main__V() {
    val bar: any = (lambda<>() = {
      val $this: any = new (constructorOf[Lhelloworld_Foo])();
      $this["bar"] = (lambda<>(arg$1: any) = {
        val prep0: int = arg$1.asInstanceOf[I];
        Lhelloworld_HelloWorld$$anon$1::bar__Lhelloworld_HelloWorld$$anon$1__I__I(this, prep0)
      });
      $this
    })();
    mod:s_Predef$.println__O__V(bar["bar"](3).asInstanceOf[I])
  }

abstract js type Lhelloworld_HelloWorld$$anon$1 extends O {
  static def bar__Lhelloworld_HelloWorld$$anon$1__I__I($this: any, x: int): int = {
    ($this.Lhelloworld_HelloWorld$$anon$1::super["bar"](x).asInstanceOf[I] *[int] 2)
    // ^- this is JSSuperBracketCall(ClassType($anon$1), $this, "bar", List(x))
  }
}
```
Now, the `JSSuperBracketCall(ClassType($anon$1), ...)` is *not* valid, because `$anon$1` is an `abstract js type`, not a `js class`. In addition, it has lost the information that it extends `Foo`, because that information was not supposed to be relevant, as that class acts purely as a repository of static defs.

There are two main ways to fix this:
* Either we eagerly desugar the `JSSuperBracketCall` in the compiler, so that it becomes something equivalent to the source code `js.constructorOf[Foo].prototype.bar.call($this, x)`, or
* We enhance the class definition of `$anon$1` with the `extends Foo`, and make it legal IR to have a `JSSuperBracketCall` of an `abstract js type`, as long as its *parent class* is a `js class`.

There are merits to both solutions. The first solution aligns better with the other treatments of anonymous JS classes, which are supposed to be a pure compiler concept, unknown of the IR. However, finding the `JSSuperBracketCall`s to replace them needs an additional pass on the generated IR (since we first emit the class as if it were named, then deconstruct the `ClassDef`).

The enhancement of the IR does not have this issue, but complicates the IR. In particular, we now need the analyzer to be able to consider an `abstract js type` as being "instantiated", so that its parent is considered "instantiated" (since the equivalent of a `LoadJSConstructor(Foo)` will eventually have to be emitted).

At this point I was about to choose the early desugaring solution and go ahead with an implementation ... and then I remembered calls to super getters and setters. And that makes the whole story much less of a clear cut.

For calls to super methods that are in fact JS getters (without `()`) or setters (ends with `_=`), we use another node, [`JSSuperBracketSelect`](https://github.com/scala-js/scala-js/blob/2350cb6a3c63430f7eb8372a538488ff1c527687/ir/src/main/scala/org/scalajs/core/ir/Trees.scala#L491). For example, for this snippet:
```scala
class Foo extends js.Object {
  var x: Int = 1

  def bar: Int = x
  def bar_=(v: Int): Unit = x = 1
}

class Bar extends Foo {
  override def bar: Int = super.bar * 2
  override def bar_=(v: Int): Unit = {
    println("bar = " + v)
    super.bar = v
  }
}
```
we have the following IR:
```scala
js class Lhelloworld_Bar extends Lhelloworld_Foo {
  def "constructor"() {
    super()
  }
  get "bar"(): any = {
    Lhelloworld_Bar::bar__Lhelloworld_Bar__I(this)
  }
  set "bar"(arg$1: any) {
    val prep0: int = arg$1.asInstanceOf[I];
    Lhelloworld_Bar::bar$und$eq__Lhelloworld_Bar__I__V(this, prep0);
    (void 0)
  }
  static def bar__Lhelloworld_Bar__I($this: any): int = {
    ($this.Lhelloworld_Bar::super["bar"].asInstanceOf[I] *[int] 2)
    // ^- this is JSSuperBracketSelect(ClassType(Bar), $this, "bar")
  }
  static def bar$und$eq__Lhelloworld_Bar__I__V($this: any, v: int) {
    mod:s_Predef$.println__O__V(("bar = " +[string] v));
    $this.Lhelloworld_Bar::super["bar"] = v
    // ^- this is Assign(JSSuperBracketSelect(ClassType(Bar), $this, "bar"), v)
  }
}
```
A non-lhs `JSSuperBracketSelect` such as the one above is eventually desugared into the following helper call:
```scala
$superGet($a_Lhelloworld_Bar().prototype, $this, "bar")
```
We see that in this case, we pass `Bar.prototype` as the first argument to [`$superGet`](https://github.com/scala-js/scala-js/blob/2350cb6a3c63430f7eb8372a538488ff1c527687/tools/scalajsenv.js#L501), called `initialProto`. (for super *calls* we used `Foo.prototype`) This value matches the concept of [*HomeObject* from the ES2015 specificiation](http://www.ecma-international.org/ecma-262/6.0/#sec-function-environment-records). In that spec, creating a "super reference" for `super.bar` starts the lookup of the property from [`envRec.GetSuperBase()`](http://www.ecma-international.org/ecma-262/6.0/#sec-getsuperbase), which basically returns `getPrototypeOf(HomeObject)`. This is what [the first step of `$resolveSuperRef`](https://github.com/scala-js/scala-js/blob/2350cb6a3c63430f7eb8372a538488ff1c527687/tools/scalajsenv.js#L490), called by `$superGet`.

Now, what if we inline `Bar` as anonymous class? We get IR like this:
```scala
  def main__V() {
    val bar: any = (lambda<>() = {
      val $this: any = new (constructorOf[Lhelloworld_Foo])();
      <linkinginfo>["envInfo"]["global"]["Object"]["defineProperty"]($this, "bar", {
        "get": (lambda<>() = {
          Lhelloworld_HelloWorld$$anon$1::bar__Lhelloworld_HelloWorld$$anon$1__I(this)
        }),
        "set": (lambda<>(arg$1: any) = {
          val prep0: int = arg$1.asInstanceOf[I];
          Lhelloworld_HelloWorld$$anon$1::bar$und$eq__Lhelloworld_HelloWorld$$anon$1__I__V(this, prep0);
          (void 0)
        }),
        "configurable": true
      });
      $this
    })();
    mod:s_Predef$.println__O__V(bar["bar"].asInstanceOf[I]);
    bar["bar"] = 4
  }

abstract js type Lhelloworld_HelloWorld$$anon$1 extends O {
  static def bar__Lhelloworld_HelloWorld$$anon$1__I($this: any): int = {
    ($this.Lhelloworld_HelloWorld$$anon$1::super["bar"].asInstanceOf[I] *[int] 2)
  }
  static def bar$und$eq__Lhelloworld_HelloWorld$$anon$1__I__V($this: any, v: int) {
    mod:s_Predef$.println__O__V(("bar = " +[string] v));
    $this.Lhelloworld_HelloWorld$$anon$1::super["bar"] = v
  }
}
```
Now again we have two "obvious" choices:
* Either desugar early in the compiler, or
* Enhance the IR to accept `JSSuperBracketSelect` with an `abstract js type`

But now, in the second case (which is supposed to be the not-best choice for super *calls*), we don't even have a `Bar.prototype` to give as a parameter to `$superGet`. In fact, the *home object* should be the `receiver` itself! (in this case, `$this`), since its `getPrototypeOf()` would correctly return `Foo.prototype`, where lookup is supposed to start.

We could therefore either change the `ClassType` into an `Option[ClassType]` in the definition of `JSSuperBracketSelect` (in which case `None` indicates that we must pass `receiver` as `initialProto`); or we could change the desugaring so that it matches more closely `JSSuperBracketCall`: we would pass `Foo.prototype` as `initialProto`, and `resolveSuperRef` would skip the first `getPrototypeOf`. The latter option seems overwhelmingly better, even though it mirrors the ES spec a bit less (`initialProto` now matches `GetSuperBase()` rather than `HomeObject`).

What if we want to desugar early in the compiler? Well ... we'll need to generate a call to `$superGet` from the compiler back-end. But that's not possible, because `$superGet` is a helper, and the IR does not know helpers (it used to, but we got rid of that for the better a long time ago). We would therefore need to add copies of `$superGet` and `$superSet` in `scalajs.runtime`, to be called by the compiler desugaring (and we still need the helper versions for non-anonymous JS classes).

In this case, it seems the enhance-the-IR solution is slightly better than the desugaring solution.

IMO we should use the same solution for super calls and super selects. But for super calls, the best solution would be desugaring, and for super selects it would be IR enhancements.

And that's where I'm currently stuck. Opinions?

**Comment by gzm0** [comment]
Fixed in ec1489927e47a2e4701bea68fbf21052aefc59ac.
